src/graphene.cpp

```diff
@@ -0,0 +1,1465 @@
+    std::set<uint64_t> setHashesToRequest;
+
+    bool fMerkleRootCorrect = true;
+    {
+        // Do the orphans first before taking the mempool.cs lock, so that we maintain correct locking order.
+        LOCK(orphanpool.cs);
+        for (auto &kv : orphanpool.mapOrphanTransactions)
+        {
+            uint64_t cheapHash = kv.first.GetCheapHash();
+
+            if (mapPartialTxHash.count(cheapHash)) // Check for collisions
+                collision = true;
+
+            mapPartialTxHash[cheapHash] = kv.first;
+        }
+
+        // We don't have to keep the lock on mempool.cs here to do mempool.queryHashes
+        // but we take the lock anyway so we don't have to re-lock again later.
+        LOCK(cs_xval);
+        mempool.queryHashes(memPoolHashes);
+
+        for (const uint256 &hash : memPoolHashes)
+        {
+            uint64_t cheapHash = hash.GetCheapHash();
+
+            if (mapPartialTxHash.count(cheapHash)) // Check for collisions
+                collision = true;
+
+            mapPartialTxHash[cheapHash] = hash;
+        }
+
+        // Add full transactions included in the block
+        for (auto &tx : vAdditionalTxs)
+        {
+            const uint256 &hash = tx->GetHash();
+            uint64_t cheapHash = hash.GetCheapHash();
+
+            if (mapPartialTxHash.count(cheapHash)) // Check for collisions
+                collision = true;
+
+            mapPartialTxHash[cheapHash] = hash;
+        }
+
+        if (!collision)
+        {
+            std::vector<uint256> localHashes;
+            for (const std::pair<uint64_t, uint256> &kv : mapPartialTxHash)
+                localHashes.push_back(kv.second);
+
+            try
+            {
+                std::vector<uint64_t> blockCheapHashes = pGrapheneSet->Reconcile(localHashes);
+
+                // Sort out what hashes we have from the complete set of cheapHashes
+                uint64_t nGrapheneTxsPossessed = 0;
+                for (size_t i = 0; i < blockCheapHashes.size(); i++)
+                {
+                    uint64_t cheapHash = blockCheapHashes[i];
+
+                    // Update mapHashOrderIndex so it is available if we later receive missing txs
+                    pfrom->grapheneMapHashOrderIndex[cheapHash] = i;
+
+                    if (mapPartialTxHash.count(cheapHash) > 0)
+                    {
+                        pfrom->grapheneBlockHashes.push_back(mapPartialTxHash[cheapHash]);
+
+                        nGrapheneTxsPossessed++;
+                    }
+                    else
+                    {
+                        pfrom->grapheneBlockHashes.push_back(nullhash);
+                        setHashesToRequest.insert(cheapHash);
+                    }
+                }
+
+                graphenedata.AddGrapheneBlockBytes(nGrapheneTxsPossessed * sizeof(uint64_t), pfrom);
+            }
+            catch (const std::runtime_error &e)
+            {
+                RequestFailoverBlock(pfrom, header.GetHash());
```

Comment:
RequestFailoverBlock takes cs_main, which causes a potential deadlock with both orphanpool.cs and cs_xval which were taken earlier in this routine.  Is it possible to release those locks earlier or to move this catch outside of the block that takes those crit sections?

recompile with "make clean; ./configure --enable-debug; make -j8" to build a debug version which will catch lock order problems.

Reply:
Given the comments from @ptschip below, does this issue require a code change? I also don't see the cs_main lock being taken, but perhaps it's nested in one of the methods I'm calling in RequestFailoverBlock?

Also, I tried compiling in debug mode as you described above. Should I see lock warnings during compilation? Right now my compilation is so noisy with warnings that I'm afraid I wouldn't notice any lock warnings go by.